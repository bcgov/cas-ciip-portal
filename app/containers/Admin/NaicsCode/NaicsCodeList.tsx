import React from 'react';
import {ListGroup} from 'react-bootstrap';
import {createFragmentContainer, graphql} from 'react-relay';
import {NaicsCodeList_query} from '__generated__/NaicsCodeList_query.graphql';

interface Props {
  query: NaicsCodeList_query;
}

export const NaicsCodeListContainer: React.FunctionComponent<Props> = ({
  query
}) => {
  return (
    <ListGroup>
      {query.allNaicsCodes.edges.map((edge) => (
        <ListGroup.Item>{edge.node.naicsCode}</ListGroup.Item>
      ))}
    </ListGroup>
  );
};

export default createFragmentContainer(NaicsCodeListContainer, {
  query: graphql`
    fragment NaicsCodeList_query on Query {
      allNaicsCodes {
        edges {
          node {
            naicsCode
            ciipSector
            naicsDescription
            deletedAt
          }
        }
      }
    }
  `
});
