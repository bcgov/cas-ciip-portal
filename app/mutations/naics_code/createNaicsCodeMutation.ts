import { graphql, DeclarativeMutationConfig } from "react-relay";
import RelayModernEnvironment from "relay-runtime/lib/store/RelayModernEnvironment";
import {
  createNaicsCodeMutationVariables,
  createNaicsCodeMutation as createNaicsCodeMutationType,
} from "createNaicsCodeMutation.graphql";
import BaseMutation from "mutations/BaseMutation";

// The default createNaicsCode mutation has been overridden by a custom function in our database in order to upsert on conflict
const mutation = graphql`
  mutation createNaicsCodeMutation($input: CreateNaicsCodeInput!) {
    createNaicsCode(input: $input) {
      naicsCodeEdge {
        node {
          id
          naicsCode
          ciipSector
          naicsDescription
          createdAt
          createdBy
          updatedAt
          updatedBy
          deletedAt
          deletedBy
        }
      }
    }
  }
`;

const createNaicsCodeMutation = async (
  environment: RelayModernEnvironment,
  variables: createNaicsCodeMutationVariables
) => {
  const connectionKey = "NaicsCodeTableContainer_allNaicsCodes";
  const configs: DeclarativeMutationConfig[] = [
    {
      type: "RANGE_ADD",
      parentID: "query",
      connectionInfo: [
        {
          key: connectionKey,
          rangeBehavior: "prepend",
        },
      ],
      edgeName: "naicsCodeEdge",
    },
  ];

  const m = new BaseMutation<createNaicsCodeMutationType>(
    "create-naics_code-mutation",
    configs
  );
  return m.performMutation(environment, mutation, variables);
};

export default createNaicsCodeMutation;
