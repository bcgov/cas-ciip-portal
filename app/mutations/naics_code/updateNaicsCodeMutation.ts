import {graphql} from 'react-relay';
import RelayModernEnvironment from 'relay-runtime/lib/store/RelayModernEnvironment';
import {
  updateNaicsCodeMutationVariables,
  updateNaicsCodeMutation as updateNaicsCodeMutationType
} from 'updateNaicsCodeMutation.graphql';
import BaseMutation from 'mutations/BaseMutation';

const mutation = graphql`
  mutation updateNaicsCodeMutation($input: UpdateNaicsCodeInput!) {
    updateNaicsCode(input: $input) {
      naicsCode {
        ...NaicsCodeTableRow_naicsCode
        deletedAt
      }
      query {
        ...NaicsCodeTable_query
      }
    }
  }
`;

const updateNaicsCodeMutation = async (
  environment: RelayModernEnvironment,
  variables: updateNaicsCodeMutationVariables
) => {
  const optimisticResponse = {
    updateNaicsCode: {
      naicsCode: {
        id: variables.input.id,
        ...variables.input.naicsCodePatch
      }
    }
  };
  const m = new BaseMutation<updateNaicsCodeMutationType>(
    'update-review-comment-mutation'
  );
  return m.performMutation(
    environment,
    mutation,
    variables,
    optimisticResponse
  );
};

export default updateNaicsCodeMutation;
