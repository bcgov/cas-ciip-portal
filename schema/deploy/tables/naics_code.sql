-- Deploy ggircs-portal:tables/naics_code to pg
-- requires: schema_ggircs_portal

begin;

create table ggircs_portal.naics_code (
  id integer primary key generated always as identity,
  naics_code varchar(1000) not null,
  ciip_sector varchar(1000),
  naics_description varchar(10000) not null
);

create unique index naics_naics_code_uindex on ggircs_portal.naics_code(naics_code);

select ggircs_portal_private.upsert_timestamp_columns(
  table_schema_name := 'ggircs_portal',
  table_name := 'naics_code',
  add_create := true,
  add_update := true,
  add_delete := true);

do
$grant$
begin
-- Grant ciip_administrator permissions
perform ggircs_portal_private.grant_permissions('select', 'naics_code', 'ciip_administrator');
perform ggircs_portal_private.grant_permissions('insert', 'naics_code', 'ciip_administrator');
perform ggircs_portal_private.grant_permissions('update', 'naics_code', 'ciip_administrator');

-- Grant ciip_analyst permissions
perform ggircs_portal_private.grant_permissions('select', 'naics_code', 'ciip_analyst');

-- Grant ciip_analyst permissions
perform ggircs_portal_private.grant_permissions('select', 'naics_code', 'ciip_industry_user');

end
$grant$;

comment on table ggircs_portal.naics_code is 'Table contains naics codes & their descriptions. NAICS is an acronym for North American Industry Classification System and is used to categorize industrial operations into sectors.';
comment on column ggircs_portal.naics_code.id is 'Primary key for naics_code table';
comment on column ggircs_portal.naics_code.naics_code is 'The naics_code, NAICS is an acronym for North American Industry Classification System and is used to categorize industrial operations into sectors.';
comment on column ggircs_portal.naics_code.ciip_sector is 'The sector this naics code belongs to according to the CleanBC Industrial Incentive Program';
comment on column ggircs_portal.naics_code.naics_description is 'The plain text description of the NAICS classification';
comment on column ggircs_portal.naics_code.created_at is 'The timestamp of when this naics_code row was created';
comment on column ggircs_portal.naics_code.created_by is 'The id of the user who created this row, references ciip_user(id)';
comment on column ggircs_portal.naics_code.updated_at is 'The timestamp of when this naics_code row was last updated';
comment on column ggircs_portal.naics_code.updated_by is 'The id of the user who last updated this row, references ciip_user(id)';;
comment on column ggircs_portal.naics_code.deleted_at is 'The timestamp of when this naics_code row was archived';
comment on column ggircs_portal.naics_code.deleted_by is 'The id of the user who archived this row, references ciip_user(id)';

commit;
