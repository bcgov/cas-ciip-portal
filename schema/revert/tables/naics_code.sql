-- Revert ggircs-portal:tables/naics_code from pg

begin;

drop table ggircs_portal.naics_code;

commit;
