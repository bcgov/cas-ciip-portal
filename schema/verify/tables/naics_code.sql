-- Verify ggircs-portal:tables/naics_code on pg

begin;

select pg_catalog.has_table_privilege('ggircs_portal.naics_code', 'select');

rollback;
