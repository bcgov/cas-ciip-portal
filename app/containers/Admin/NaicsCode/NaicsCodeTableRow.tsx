import React from 'react';
import {Button} from 'react-bootstrap';
import {graphql, createFragmentContainer, RelayProp} from 'react-relay';
import {NaicsCodeTableRow_naicsCode} from '__generated__/NaicsCodeTableRow_naicsCode.graphql';
import updateNaicsCodeMutation from 'mutations/naics_code/updateNaicsCodeMutation';
import {nowMoment} from 'functions/formatDates';

interface Props {
  relay: RelayProp;
  naicsCode: NaicsCodeTableRow_naicsCode;
  connectionId: string;
}

export const NaicsCodeTableRowContainer: React.FunctionComponent<Props> = (
  props
) => {
  const {naicsCode, connectionId} = props;

  const handleDeleteNaicsCode = async () => {
    const {environment} = props.relay;
    const variables = {
      input: {
        id: naicsCode.id,
        naicsCodePatch: {
          naicsCode: naicsCode.naicsCode,
          ciipSector: naicsCode.ciipSector,
          naicsDescription: naicsCode.naicsDescription,
          deletedAt: nowMoment().format('YYYY-MM-DDTHH:mm:ss')
        }
      }
    };
    await updateNaicsCodeMutation(environment, variables, connectionId);
  };

  return (
    <tr key={naicsCode.id}>
      <td>{naicsCode.naicsCode}</td>
      <td>{naicsCode.ciipSector}</td>
      <td>{naicsCode.naicsDescription}</td>
      <td style={{textAlign: 'center'}}>
        <Button variant="danger" onClick={() => handleDeleteNaicsCode()}>
          Delete
        </Button>
      </td>
    </tr>
  );
};

export default createFragmentContainer(NaicsCodeTableRowContainer, {
  naicsCode: graphql`
    fragment NaicsCodeTableRow_naicsCode on NaicsCode {
      id
      naicsCode
      ciipSector
      naicsDescription
      deletedAt
    }
  `
});
