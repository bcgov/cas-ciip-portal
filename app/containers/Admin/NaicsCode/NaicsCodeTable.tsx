import React, { useState } from "react";
import { Table, Button } from "react-bootstrap";
import { graphql, createFragmentContainer, RelayProp } from "react-relay";
import { NaicsCodeTable_query } from "__generated__/NaicsCodeTable_query.graphql";
import NaicsCodeTableRow from "./NaicsCodeTableRow";
import CreateNaicsCodeModal from "components/Admin/CreateNaicsCodeModal";
import createNaicsCodeMutation from "mutations/naics_code/createNaicsCodeMutation";
import withPromiseLoading from "lib/withPromiseLoading";

const LoadingCreateNaicsCodeModal = withPromiseLoading(
  CreateNaicsCodeModal,
  "onSubmit",
  "disabled"
);
interface Props {
  relay: RelayProp;
  query: NaicsCodeTable_query;
}

export const NaicsCodeTableContainer: React.FunctionComponent<Props> = (
  props
) => {
  const [validated, setValidated] = useState(false);
  const [showCreateModal, setShowCreateModal] = useState(false);
  const [showActiveCodeError, setShowActiveCodeError] = useState(false);
  const { query } = props;

  const handleClose = () => {
    setShowCreateModal(false);
    setValidated(false);
    setShowActiveCodeError(false);
  };

  const naicsCodeIsActive = (naicsCode) => {
    return query.allNaicsCodes.edges.some(
      (edge) => edge.node.naicsCode === naicsCode
    );
  };

  const handleCreateNaicsCode = async (form: any) => {
    setValidated(true);

    if (naicsCodeIsActive(form[1].value)) setShowActiveCodeError(true);
    else if (form.checkValidity() === true) {
      const { environment } = props.relay;
      const variables = {
        input: {
          naicsCodeInput: form[1].value,
          ciipSectorInput: form[2].value ? form[2].value : null,
          naicsDescriptionInput: form[3].value,
        },
      };

      await createNaicsCodeMutation(environment, variables);
    }
  };

  return (
    <>
      <div style={{ textAlign: "right" }}>
        <Button
          style={{ marginTop: "-100px" }}
          onClick={() => setShowCreateModal(true)}
        >
          New NAICS Code
        </Button>
      </div>
      <LoadingCreateNaicsCodeModal
        validated={validated}
        onSubmit={handleCreateNaicsCode}
        show={showCreateModal}
        onClose={() => handleClose()}
        showActiveCodeError={showActiveCodeError}
      />
      <Table striped bordered hover>
        <thead>
          <tr>
            <th scope="col">NAICS Code</th>
            <th scope="col">CIIP Sector</th>
            <th scope="col">NAICS Description</th>
            <th />
          </tr>
        </thead>
        <tbody>
          {query.allNaicsCodes?.edges.map(({ node }) => {
            return (
              <NaicsCodeTableRow
                key={node.id}
                naicsCode={node}
                connectionId={query.allNaicsCodes.__id}
              />
            );
          })}
        </tbody>
      </Table>
      <style jsx global>{`
        .table thead th {
          color: white;
          background: #003366;
        }
      `}</style>
    </>
  );
};

export default createFragmentContainer(NaicsCodeTableContainer, {
  query: graphql`
    fragment NaicsCodeTable_query on Query {
      allNaicsCodes(
        first: 2147483647
        filter: { deletedAt: { isNull: true } }
        orderBy: NAICS_CODE_ASC
      ) @connection(key: "NaicsCodeTableContainer_allNaicsCodes", filters: []) {
        __id
        edges {
          node {
            id
            naicsCode
            ...NaicsCodeTableRow_naicsCode
          }
        }
      }
    }
  `,
});
