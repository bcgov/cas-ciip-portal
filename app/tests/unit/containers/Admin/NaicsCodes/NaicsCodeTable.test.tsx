import React from 'react';
import {shallow} from 'enzyme';
import {NaicsCodeTableContainer} from 'containers/Admin/NaicsCode/NaicsCodeTable';

describe('NaicsCodeTable', () => {
  it('should match the snapshot with the NaicsCodeTable container', async () => {
    const renderer = shallow(
      <NaicsCodeTableContainer
        relay={null}
        query={{
          ' $refType': 'NaicsCodeTable_query',
          allNaicsCodes: {
            __id: 'abc',
            edges: [
              {
                node: {
                  id: 'abc',
                  ' $fragmentRefs': {
                    NaicsCodeTableRow_naicsCode: true
                  }
                }
              }
            ]
          }
        }}
      />
    );
    expect(renderer).toMatchSnapshot();
  });
  it('should render the NaicsCodeTableRow container', async () => {
    const renderer = shallow(
      <NaicsCodeTableContainer
        relay={null}
        query={{
          ' $refType': 'NaicsCodeTable_query',
          allNaicsCodes: {
            __id: 'abc',
            edges: [
              {
                node: {
                  id: 'abc',
                  ' $fragmentRefs': {
                    NaicsCodeTableRow_naicsCode: true
                  }
                }
              }
            ]
          }
        }}
      />
    );
    expect(renderer.exists('Relay(NaicsCodeTableRowContainer)')).toBe(true);
  });

  it('should render the CreateNaicsCode modal component', async () => {
    const renderer = shallow(
      <NaicsCodeTableContainer
        relay={null}
        query={{
          ' $refType': 'NaicsCodeTable_query',
          allNaicsCodes: {
            __id: 'abc',
            edges: [
              {
                node: {
                  id: 'abc',
                  ' $fragmentRefs': {
                    NaicsCodeTableRow_naicsCode: true
                  }
                }
              }
            ]
          }
        }}
      />
    );
    expect(renderer.exists('CreateNaicsCodeModal')).toBe(true);
  });

  it('should open CreateNaicsCode modal component on New NAICS Code Button click', async () => {
    const renderer = shallow(
      <NaicsCodeTableContainer
        query={{
          ' $refType': 'NaicsCodeTable_query',
          allNaicsCodes: {
            __id: 'abc',
            edges: [
              {
                node: {
                  id: 'abc',
                  ' $fragmentRefs': {
                    NaicsCodeTableRow_naicsCode: true
                  }
                }
              }
            ]
          }
        }}
        relay={null}
      />
    );

    renderer.find('Button').first().simulate('click');

    expect(renderer.find('CreateNaicsCodeModal').prop('show')).toBe(true);
  });

  it('should call the create mutation on Form submit', async () => {
    const spy = jest.spyOn(
      require('mutations/naics_code/createNaicsCodeMutation'),
      'default'
    );
    spy.mockImplementation(() => null);
    const relay = {environment: null};
    const renderer = shallow(
      <NaicsCodeTableContainer
        query={{
          ' $refType': 'NaicsCodeTable_query',
          allNaicsCodes: {
            __id: 'abc',
            edges: [
              {
                node: {
                  id: 'abc',
                  ' $fragmentRefs': {
                    NaicsCodeTableRow_naicsCode: true
                  }
                }
              }
            ]
          }
        }}
        relay={relay as any}
      />
    );

    renderer.find('CreateNaicsCodeModal').prop('onSubmit')({
      stopPropagation: jest.fn(),
      preventDefault: jest.fn(),
      persist: jest.fn(),
      currentTarget: {checkValidity: jest.fn(() => true)},
      target: [{value: 'a'}, {value: 'b'}, {value: 'c'}]
    } as any);

    expect(spy).toBeCalledTimes(1);
    expect(spy).toBeCalledWith(null, {
      input: {
        ciipSectorInput: 'b',
        naicsCodeInput: 'a',
        naicsDescriptionInput: 'c'
      }
    });
  });
});
