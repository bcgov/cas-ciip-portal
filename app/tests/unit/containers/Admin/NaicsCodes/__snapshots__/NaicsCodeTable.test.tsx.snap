// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NaicsCodeTable should match the snapshot with the NaicsCodeTable container 1`] = `
<Fragment>
  <div
    className="jsx-1364320900"
    style={
      Object {
        "textAlign": "right",
      }
    }
  >
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginTop": "-100px",
        }
      }
      variant="primary"
    >
      New Naics Code
    </Button>
  </div>
  <CreateNaicsCodeModal
    onClose={[Function]}
    onSubmit={[Function]}
    show={false}
    validated={false}
  />
  <ForwardRef
    bordered={true}
    hover={true}
    striped={true}
  >
    <thead
      className="jsx-1364320900"
    >
      <tr
        className="jsx-1364320900"
      >
        <th
          className="jsx-1364320900"
          scope="col"
        >
          NAICS Code
        </th>
        <th
          className="jsx-1364320900"
          scope="col"
        >
          CIIP Sector
        </th>
        <th
          className="jsx-1364320900"
          scope="col"
        >
          NAICS Description
        </th>
        <th
          className="jsx-1364320900"
        />
      </tr>
    </thead>
    <tbody
      className="jsx-1364320900"
    >
      <Relay(NaicsCodeTableRowContainer)
        connectionId="abc"
        key="abc"
        naicsCode={
          Object {
            " $fragmentRefs": Object {
              "NaicsCodeTableRow_naicsCode": true,
            },
            "id": "abc",
          }
        }
      />
    </tbody>
  </ForwardRef>
  <JSXStyle
    id="1364320900"
  >
    .table thead th{color:white;background:#003366;}
  </JSXStyle>
</Fragment>
`;
